# oriGen â€“ Analysis Code

- The code in this repository are meant to reproduce results from: TO-DO-add url and doi after paper is published

Workflows for the oriGen Analysis of 1,427 whole genomes.  

Welcome! ğŸ‘‹  
This repository contains code used to analyze the data in the oriGen study, which involved whole-genome sequencing of 1,427 individuals. Our goal is to provide transparent and reproducible access to the workflows behind our results.

## ğŸ“„ Contents of this repository  

The oriGen study is a population-scale genomic project focused on Mexican individuals. This release includes the following subdirectories:  

- **oGpv3**: NGS preprocessing, QC, alignment and Variant Calling
- **vep-extended**: Variant Effect Annotation
- **LoFbySample**: Loss of Function variation summary
- **vcf2kinship**: Kinship Analysis with KING2
- **vcf2pcp**: ADMIXTURE and Principal Component Analysis from joint population VCF
- **cnv-analysis**: Processing of the CNV BCFs from CNVpytor
- **cov-calculation**: NGS coverage calculation by exon for CN estimation
- **viral-detection**: Detection of viral hits in sequencing data

## How to use this repository

Each workflow subdirectory contains:

* The Source code used to run analysis on a part of the paper
* A small (simulated or real) dataset to demo the code
* All software dependencies and operating systems  
* Installation instructions
* Output descriptions
* Expected runtimes

### * **No sensitive nor personal data is shared in the repositories** *

Every subdirectory includes a ./scripts/ dir with particular R, python or bash scripts for descriptive and or statistical analyses.

Every subdirectory includes a main.nf, and a modules/ subdir with the nextflow framework to run.

## ğŸ“ Workflow Subdirectory Structure

```bash
â”œâ”€â”€ cnv-analysis/    # CNV Analysis from CNVpytor output
â”œâ”€â”€ cov-calculation  # Coverage calculation for CNV estimation
â”œâ”€â”€ LoFbySample/     # Loss of Function Summary
â”œâ”€â”€ oGpv3            # pipeline for pre-processing, QC, alignment and short variant calling
â”œâ”€â”€ vcf2kinship/     # Kinship Analysis
â”œâ”€â”€ vcf2pcp/         # ADMIXTURE and Principal Component Analysis
â”œâ”€â”€ vep-extended/    # Variant Effect Annotation
â”œâ”€â”€ viral-detection  # Analysis of viral hits in NGS data
â””â”€â”€ README.md        # This file
```

### Contact
If you have questions, requests, or bugs to report, please open an issue in this github page.  

### Dev Team
Israel Aguilar-Ordonez <iaguilaror@gmail.com>   
Victor Trevino Alvarado <vtrevino@tec.mx>   
Eugenio Guzman Cerezo <eugenio.guzman@tec.mx>   

This code was developed as part of Israel Aguilar-OrdoÃ±ezâ€™s postdoctoral research at TecnolÃ³gico de Monterrey during the 2024â€“2025.

### Cite us

If you find the code in this repository useful, please include the following citation in your work:

TO-DO after publication.  
